File: Boggle.hs
module Boggle (boggle) where
import Data.Map (Map)
import Data.List ()
import qualified Data.Map as Map

{--
    Fill in the boggle function below. Use as many helpers as you want.
    Test your code by running 'cabal test' from the tester_hs_simple directory.
--}

    
boggle :: [String] -> [String] -> [ (String, [ (Int, Int) ] ) ]
boggle board words = Map.toList ans
    where
        trie = createTrie words Map.empty
        wordsMap = wordsToMap words Map.empty
        ans = runner board (0,0) trie Map.empty (length board - 1) wordsMap

test :: Map.Map String [(Int, Int)]
test = ans
    where
        trie = createTrie ["as","at","ate","eat","set","sat","eats","east","seat","sate"] Map.empty
        board = ["ea","st"]
        visited = []
        currentWord = ""
        words = wordsToMap ["as","at","ate","eat","set","sat","eats","east","seat","sate"] Map.empty
        wordsToReturn = Map.empty
        ans = runner board (0,0) trie Map.empty (length board - 1) words


runner :: [String] -> (Int, Int) -> Map.Map String Int -> Map.Map String [(Int,Int)] -> Int -> Map.Map String Int -> Map.Map String [(Int,Int)]
runner board (row, col) trie wordsToReturn maxLen words
    | (row == maxLen) && (col == maxLen) = dfs board (maxLen, maxLen) [] wordsToReturn trie "" words
    | col == maxLen = runner board (row + 1, 0) trie o maxLen words
    | otherwise = runner board (row, col + 1) trie o maxLen words
        where 
            o = dfs board (row, col) [] wordsToReturn trie "" words

wordsToMap :: [String] -> Map.Map String Int -> Map.Map String Int
wordsToMap [] map = map
wordsToMap (word : rest) map = wordsToMap rest newMap
    where 
        newMap = Map.insert word 1 map

dfs :: [String] -> (Int, Int) -> [(Int, Int)] -> Map.Map String [(Int, Int)] -> Map.Map String Int  -> String -> Map.Map String Int -> Map.Map String [(Int, Int)]
dfs board (y, x) visited wordsToReturn trie currentWord words
    | Map.member currentWord trie && notElem (y, x) visited = c
    | otherwise = wordsToReturn 
    where 
        movesList = [(1,0),(-1,0),(0,1),(0,-1),(-1,-1),(1,-1),(-1,1),(1,1)]
        newLetter = (board !! y) !! x
        newCurrentWord = currentWord ++ [newLetter]
        newVisited = (y, x) : visited 
        newWordsToReturn = if Map.member newCurrentWord words then Map.insert newCurrentWord (reverse newVisited) wordsToReturn else wordsToReturn
        nextMoves = possibleMoves (y,x) movesList (length board) []
        c = if not (Map.member newCurrentWord trie) then wordsToReturn else foldl (\acc (row, col) -> dfs board (row, col) newVisited acc trie newCurrentWord words) newWordsToReturn nextMoves


possibleMoves :: (Int, Int) -> [(Int, Int)] -> Int -> [(Int, Int)] -> [(Int, Int)]
possibleMoves (y, x) [] fullLength possibleMovesList =  possibleMovesList
possibleMoves (y, x) ((y2, x2) : rest) fullLength possibleMovesList = possibleMoves (y, x) rest fullLength newPossibleMovesList
    where 
        newPossibleMovesList = 
            if y + y2 <= fullLength - 1 && y + y2 >= 0 && x + x2 <= fullLength - 1 && x + x2 >= 0 
                then  (y + y2, x + x2) : possibleMovesList
                else possibleMovesList
        

createTrie :: [String] -> Map.Map String Int -> Map.Map String Int 
createTrie [] dict = dict 
createTrie (word : rest ) dict = createTrie rest updatedDict 
    where 
        updatedDict = addWordToDict word dict 0 


addWordToDict :: String -> Map.Map String Int -> Int -> Map.Map String Int
addWordToDict s d l
    | l == length s = Map.insert (take l s) 1 d 
    | otherwise = addWordToDict s d2 l2
        where 
            d2 = Map.insert (take l s) 1 d 
            l2 = l + 1

File: boggle.ex
defmodule Boggle do

  @moduledoc """
    Add your boggle function below. You may add additional helper functions if you desire.
    Test your code by running 'mix test' from the tester_ex_simple directory.
  """

  def boggle(board, words) do
    len = tuple_size(board)
    # words = Enum.filter(words, fn str -> String.length(str) <= (len * len) end)
    trie = createTrie(words, len)
    len = tuple_size(board)
    m = runner(board, trie, len)
    finalAnswer(words, m, len)
  end


  def finalAnswer(words, m, len) do
    finalAnswer(words, m, %{}, len)
  end

  def finalAnswer([], _m, result, _max) do
    result
  end

  def finalAnswer([h | t], m, result, max) do

    if Map.has_key?(m,h) do
      finalAnswer(t, m , Map.put(result, h, Map.get(m, h)), max)
    else
      finalAnswer(t, m , result, max)
    end

  end

  def runner(board, trie, len) do
    runner(0,0, board, trie, %{}, len - 1)
  end

  def runner(len, len, board, trie, m, len) do
    dfs(board, {len, len}, [], m, "", trie)
  end

  def runner(row, len, board, trie, m, len) do
    r = dfs(board, {row, len}, [], m, "", trie)
    runner(row + 1,0, board, trie, r, len)
  end

  def runner(row, col, board, trie, m, len) when col < (len) do
    r = dfs(board, {row, col}, [], m, "", trie)
    runner(row,col + 1, board, trie, r , len)
  end

  def dfs(matrix, trie, startPosition) do
    dfs(matrix, startPosition, [], %{}, "", trie)
  end

  def dfs(matrix, currentPosition, visitedList, wordsToReturn, currentWord, trie) do
    if Map.has_key?(trie,currentWord) do

      if Enum.member?(visitedList, currentPosition) do

        wordsToReturn
      else

        col = elem(currentPosition, 1)
        row = elem(currentPosition, 0)

        len = tuple_size(matrix)

        newLetter = elem(elem(matrix, row), col)

        visitedList = visitedList ++ [currentPosition]

        currentWord = currentWord <> newLetter

        wordsToReturn = Map.put(wordsToReturn, currentWord, visitedList)

        possibleMoves = possibleMoves(currentPosition, len)

        #IO.puts "visited #{inspect visitedList} currentWord #{inspect currentWord} wordsToReturn #{inspect wordsToReturn} trie #{inspect trie}"

        c = Enum.reduce(possibleMoves, wordsToReturn, fn {row, col}, acc ->
          dfs(matrix, {row, col}, visitedList, acc, currentWord, trie)
        end)
        c
      end

    else
      wordsToReturn
    end

  end


  def possibleMoves(currentPosition, len) do
    moves = [{1,0},{-1,0},{0,1},{0,-1},{-1,-1},{1,-1},{-1,1},{1,1}]
    possibleMoves(currentPosition, moves, len - 1, [])
  end

  def possibleMoves(_currentPosition,[], _len, acc) do
    acc
  end

  def possibleMoves(currentPosition,[h | t], len, acc) do
    col = elem(currentPosition, 1)
    row = elem(currentPosition, 0)
    colt = elem(h, 1)
    rowt = elem(h, 0)
    if (col + colt) >= 0 and (col + colt) <= len and (row + rowt) >= 0 and (row + rowt) <= len do
      possibleMoves(currentPosition, t, len, [{row + rowt, col + colt} | acc])
    else
      possibleMoves(currentPosition, t, len, acc)
    end
  end

  def createTrie(words, len) do
    createTrie(words, %{"" => 1}, len)
  end

  def createTrie([h | t], m, len) do
    if String.length(h) > (len * len) do
      createTrie(t,m, len)
    else
      m = addWordToDict(h, m)
      createTrie(t, m, len)
    end

  end

  def createTrie([], m, _len) do
    m
  end

  def addWordToDict(word, d) do
    addWordToDict(word, d, 1, String.length(word))
  end

  def addWordToDict(word, d, l, l) do
    Map.put(d,String.slice(word,0, l),1)
  end

  def addWordToDict(word, d, s, l) do
    d = Map.put(d,String.slice(word,0,s),1)
    addWordToDict(word, d, s + 1, l)
  end

end
